### ISSUE TITLE
Use HMAC-SHA1 instead of simple concatenation

### ISSUE THREAD

**Issue by gsauthof** [thilomaurer/pwcalc#6]
As far as I can see, the generated password string is just the SHA1 hash of the concatenation of the secret and the alias (domain):

> The formula is as simple as "[secret][alias]" → SHA1 → BASE64 

(README.md)

And the source code, [extension.js, calculatePassword()](https://github.com/thilomaurer/pwcalc/blob/master/extension.js#L250):

```
function calculatePassword(secret, domain) {
        if (secret==""||domain=="") return "";
        var sha1 = Utils.Sha1.hash(secret + domain);
        var array = sha1.match(/.{1,2}/g);
        var bytes = new Uint8Array(20);
        for (var i = 0, j = array.length; i < j; i += 1)
            bytes[i] = parseInt(array[i], 16);
        var base64 = Base64.base64.encode(String.fromCharCode.apply(null, bytes));
        return base64.substring(0, 16);
}
```

Thus, the generated password is vulnerable to length-extension attacks. See also:

https://en.wikipedia.org/wiki/Hash-based_message_authentication_code#Design_principles

Just using HMAC-SHA1 would resolve this issue.

**Comment by thilomaurer** [comment]
While it is true that MAC-SHA1 is vulnerate to length-extension attacks, the user's secret is nowhere in jeopardy.

The vulnerability is that an attacker can generate passwords for aliases that are not relevant to the user, for example (in analogy to https://blog.skullsecurity.org/2012/everything-you-need-to-know-about-hash-length-extension-attacks), an attacker hacking facebook or your device and thereby receiving the full hash for your user alias "facebook.com", will also be able to generate a password for each alias similar to 
```
facebook.com\x80\x00\x00
\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00
\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00
\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00
\x00\x00\x02\x28"+(any string)
```
which would be of no significant use for an attacker. Do you agree?

Nevertheless I plan to add a HMAC-SHA1 feature to the password calculator and set it to default for new users. Pull requests would be welcomed.

**Comment by fortaa** [comment]
try this one.
```javascript

function calculatePassword(secret, domain, length) {
    var i_key_b = new Uint8Array(64), o_key_b = new Uint8Array(64);
    var i, j;
    
    if (secret==""||domain=="") return "";

    secret = Utils.Utf8.encode(secret);
    domain = Utils.Utf8.encode(domain);

    if (secret.length > 64) {
        secret = Utils.Sha1.hash(secret, false);
    }

    i_key_b.fill(0x36);
    o_key_b.fill(0x5c);
    
    for (i = 0, j = secret.length; i < j; i += 1) {
        i_key_b[i] ^= secret.charCodeAt(i);
        o_key_b[i] ^= secret.charCodeAt(i);
    }

    sha1 = Utils.Sha1.hash(String.fromCharCode.apply(null, i_key_b) + domain, false);
    sha1 = Utils.Sha1.hash(String.fromCharCode.apply(null, o_key_b) + sha1, false);
    
    var array = sha1.match(/.{2}/g);    
    var bytes = new Uint8Array(20);
    for (i = 0, j = array.length; i < j; i += 1)
        bytes[i] = parseInt(array[i], 16);
    var base64 = Base64.base64.encode(String.fromCharCode.apply(null, bytes));
    return base64.substring(0, length);
}
```

**Comment by thilomaurer** [comment]
Implemented with e44a87b60d8d1dc88f2cbd6cf95c3d7fb70c65a6
Awaiting review by Gnome extensions team.
